packages/mux-uploader/src/index.ts

```diff
@@ -0,0 +1,419 @@
+}
+
+:host([upload-in-progress]) ::slotted(p) {
+  display: block;
+}
+
+:host([type="bar"][upload-error]) .progress-bar {
+  background: #e22c3e;
+}
+
+:host([type="bar"][upload-error]) .status-message {
+  color: #e22c3e;
+}
+
+:host([type="radial"][upload-error]) .status-message {
+  color: #e22c3e;
+}
+
+:host([upload-error][enableStatus]) .upload-status {
+  display: none;
+}
+
+:host([upload-error]) .retry-message {
+  display: inline-block;
+}
+
+:host([upload-error]) .text-container {
+  display: flex;
+}
+
+:host([upload-error]) ::slotted(p) {
+  display: none;
+}
+
+.upload-status {
+  font-size: inherit;
+  margin-bottom: 16px;
+}
+
+.progress-bar {
+  box-shadow: 0 10px 40px -10px #fff;
+  border-radius: 100px;
+  background: var(--progress-bar-fill-color, #000000);
+  height: 4px;
+  width: 0%;
+}
+
+:host([upload-in-progress]) button {
+  display: none;
+}
+
+:host([upload-in-progress]) ::slotted(button) {
+  display: none;
+}
+
+:host([upload-in-progress]) .upload-instruction {
+  display: none;
+}
+
+circle {
+  stroke: var(--progress-radial-fill-color, black);
+  stroke-width: 6;  /* Thickness of the circle */
+  fill: transparent; /* Make inside of the circle see-through */
+
+  /* Animation */ 
+  transition: 0.35s;
+  transform: rotate(-90deg);
+  transform-origin: 50% 50%;
+  -webkit-transform-origin: 50% 50%;
+  -moz-transform-origin: 50% 50%;
+}
+`;
+
+const template = document.createElement('template');
+
+template.innerHTML = `
+<style>
+  ${styles}
+</style>
+
+<p class="upload-instruction" id="upload-instruction">Drop file to upload</p>
+
+<div class=text-container>
+  <span class="status-message" id="status-message"></span>
+  <span class="retry-message" id="retry-message">Try again</span>
+</div>
+
+<input type="file" />
+<slot name="custom-button"><button type="button">Upload video</button></slot>
+<slot name="custom-progress"><p class="upload-status" id="upload-status"></p></slot>
+
+<div class="bar-type">
+  <div class="progress-bar" id="progress-bar"></div>
+</div>
+<div class="radial-type">
+  <svg
+    width="120"
+    height="120">
+    <!-- To prevent overflow of the SVG wrapper, radius must be  (svgWidth / 2) - (circleStrokeWidth * 2)
+      or use overflow: visible on the svg.-->
+    <circle
+      r="52"
+      cx="60"
+      cy="60"
+    />
+  <svg>
+</div>
+
+<slot name="dropzone">
+  <mux-uploader-drop></mux-uploader-drop>
+</slot>
+`;
+
+// Note: Use "bar" for now since the CSS for radial is WIP. (TD).
+const TYPES = {
+  BAR: 'bar',
+  RADIAL: 'radial',
+};
+
+class MuxUploaderElement extends HTMLElement {
+  hiddenFileInput: HTMLInputElement | null | undefined;
+  filePickerButton: HTMLButtonElement | null | undefined;
+  svgCircle: SVGCircleElement | null | undefined;
+  progressBar: HTMLElement | null | undefined;
+  uploadPercentage: HTMLElement | null | undefined;
+  statusMessage: HTMLElement | null | undefined;
+  retryMessage: HTMLElement | null | undefined;
+  _dropHandler: Function;
```

Comment:
nit (non-blocking): we started using real private props in Mux player to keeping the internals as private as possible and have a unpolluted API. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Private_class_fields

cool thing is that esbuild will automatically transform these to use WeakMaps for browsers that don't support them yet.

Reply:
TIL!